FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,50 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let newCache = try ManagedCache.newUniqueInstance(in: context)
+				newCache.timestamp = timestamp
+				newCache.feed = ManagedFeedImage.images(from: feed, in: context, with: newCache)
+
+				try context.save()
+				completion(nil)
+
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try ManagedCache.deletePreviousCache(in: context)
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+
+	private func perform(_ action: @escaping (NSManagedObjectContext) -> Void) {
+		let context = self.context
+		context.perform { action(context) }
```

Comment:
Another way of capturing the `context` inside the closure without capturing `self` is to use a capture list as shown below:

```suggestion
		context.perform { [context] in action(context) }
```

It achieves the same result with one less line of code.